File: frontend/src/types/transaction.ts
export type FlowDirection = "inflow" | "outflow";
export type TransactionStatus = "pending" | "booked" | "rejected";
export type TransactionType =
  | "credit_transfer"
  | "direct_debit"
  | "card_payment"
  | "cash"
  | "fee"
  | "interest"
  | "tax"
  | "other";
export type CategorizationSource = "manual" | "automatic" | "rule" | "import";
export type BatchStatus = "pending" | "processing" | "completed" | "failed" | "partial";

export interface Allocation {
  id: string;
  transaction_id: string;
  category_id: string;
  subcategory_id: string | null;
  amount: number;
  currency: string;
  percentage: number | null;
  category?: Category | null;
  subcategory?: Subcategory | null;
  created_at: string;
  updated_at: string;
}

export interface Transaction {
  id: string;
  company_id: string;
  bank_account_id: string;
  amount: number;
  currency: string;
  direction: FlowDirection;
  transaction_date: string;
  value_date: string | null;
  description: string | null;
  remittance_info: string | null;
  transaction_type: TransactionType;
  category_id: string | null;
  subcategory_id: string | null;
  categorization_source: CategorizationSource | null;
  counterpart_id: string | null;
  counterpart_name: string | null;
  counterpart_iban: string | null;
  status: TransactionStatus;
  verified: boolean;
  provider_transaction_id: string | null;
  hidden_at: string | null;
  notes: string | null;
  import_batch_id: string | null;
  reconciliation_status: ReconciliationStatus | null;
  reconciliation_id: string | null;
  category?: Category | null;
  subcategory?: Subcategory | null;
  allocations: Allocation[];
  bank_account_nickname: string | null;
  created_at: string;
  updated_at: string;
}

export interface TransactionListResponse {
  items: Transaction[];
  total: number;
  page: number;
  page_size: number;
}

export interface TransactionTotals {
  positive: number;
  negative: number;
}

export interface TransactionMetadata {
  total: number;
  totals: TransactionTotals;
  total_uncategorized: number;
}

export interface ImportBatch {
  id: string;
  status: BatchStatus;
  filename: string | null;
  total_records: number;
  imported_records: number;
  skipped_records: number;
  error_records: number;
  errors: Array<{ row: number; error: string }>;
  started_at: string | null;
  completed_at: string | null;
}

// Re-export for convenience
import type { Category, Subcategory } from "./category";
import type { ReconciliationStatus } from "./reconciliation";
export type { Category, Subcategory };

File: backend/app/schemas/transaction.py
import uuid
from datetime import date, datetime
from decimal import Decimal
from pydantic import BaseModel

from app.schemas.category import CategoryResponse, SubcategoryResponse


class AllocationResponse(BaseModel):
    id: uuid.UUID
    transaction_id: uuid.UUID
    category_id: uuid.UUID
    subcategory_id: uuid.UUID | None = None
    amount: Decimal
    currency: str
    percentage: Decimal | None = None
    category: CategoryResponse | None = None
    subcategory: SubcategoryResponse | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class TransactionResponse(BaseModel):
    id: uuid.UUID
    company_id: uuid.UUID
    bank_account_id: uuid.UUID
    amount: Decimal
    currency: str
    direction: str
    transaction_date: date
    value_date: date | None = None
    description: str | None = None
    remittance_info: str | None = None
    transaction_type: str
    category_id: uuid.UUID | None = None
    subcategory_id: uuid.UUID | None = None
    categorization_source: str | None = None
    counterpart_id: uuid.UUID | None = None
    counterpart_name: str | None = None
    counterpart_iban: str | None = None
    status: str
    verified: bool
    provider_transaction_id: str | None = None
    hidden_at: datetime | None = None
    notes: str | None = None
    reconciliation_status: str = "unreconciled"
    reconciliation_id: uuid.UUID | None = None
    import_batch_id: uuid.UUID | None = None
    category: CategoryResponse | None = None
    subcategory: SubcategoryResponse | None = None
    allocations: list[AllocationResponse] = []
    bank_account_nickname: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class TransactionListResponse(BaseModel):
    items: list[TransactionResponse]
    total: int
    page: int
    page_size: int


class TransactionTotals(BaseModel):
    positive: Decimal
    negative: Decimal


class TransactionMetadataResponse(BaseModel):
    total: int
    totals: TransactionTotals
    total_uncategorized: int


class UpdateTransactionRequest(BaseModel):
    category_id: uuid.UUID | None = None
    subcategory_id: uuid.UUID | None = None
    notes: str | None = None
    verified: bool | None = None
    counterpart_name: str | None = None


class CSVMappingRequest(BaseModel):
    date_column: str
    amount_column: str
    description_column: str | None = None
    date_format: str = "%d/%m/%Y"
    decimal_separator: str = ","
    csv_separator: str = ";"
